presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/SphericalGeoFunctions.java

```diff
@@ -236,6 +257,55 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
         return sum * 1000;
     }
 
+    @Description("Returns the Point value that is the mathematical centroid of a Spherical Geography")
+    @ScalarFunction("ST_Centroid")
+    @SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME)
+    public static Slice stSphericalCentroid(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) Slice input)
+    {
+        OGCGeometry geometry = EsriGeometrySerde.deserialize(input);
+        if (geometry.isEmpty()) {
+            return Slices.EMPTY_SLICE;
+        }
+        // TODO: add support for other types e.g. POLYGON
+        validateSphericalType("ST_Centroid", geometry, EnumSet.of(POINT, MULTI_POINT));
+        if (geometry instanceof OGCPoint) {
+            return input;
+        }
+
+        OGCGeometryCollection geometryCollection = (OGCGeometryCollection) geometry;
+        for (int i = 0; i < geometryCollection.numGeometries(); i++) {
+            OGCGeometry g = geometryCollection.geometryN(i);
+            validateSphericalType("ST_Centroid", g, EnumSet.of(POINT));
+            Point p = (Point) g.getEsriGeometry();
+            checkLongitude(p.getX());
+            checkLatitude(p.getY());
+        }
+
+        Point centroid;
+        if (geometryCollection.numGeometries() == 1) {
+            centroid = (Point) geometryCollection.geometryN(0).getEsriGeometry();
+        }
+        else {
+            double x3DTotal = 0;
+            double y3DTotal = 0;
+            double z3DTotal = 0;
+
+            for (int i = 0; i < geometryCollection.numGeometries(); i++) {
+                CartesianPoint cp = new CartesianPoint((Point) geometryCollection.geometryN(i).getEsriGeometry());
+                x3DTotal += cp.getX();
+                y3DTotal += cp.getY();
+                z3DTotal += cp.getZ();
+            }
+
+            double centroidVectorLength = Math.sqrt(x3DTotal * x3DTotal + y3DTotal * y3DTotal + z3DTotal * z3DTotal);
```

Comment:
Let's explicitly check for `centroidVectorLength == 0.0` here; I'm worried that some float precision is hiding a possible div-by-0 bug.

Reply:
I think this would be very unlikely as it's a sum of squares. The only way it could happen would be that all the coordinates are actually zero (center of earth). Nonetheless, I've added a check that throws an exception if it happens.

Reply:
When you have queries with 10's of billions of rows, and millions of queries per day, "very unlikely" happens surprisingly frequently!  And propagating NaN's are quite annoying to track down.